### PR TITLE
Added SNAPSHOT[0-9]*.\.jar build and publish + set version script

### PR DESCRIPTION
**Pull request by ysibirski**
First iteration of improving CI/CD pipeline as we as adding set_version script.

### CODE REVIEW

File: api/Cargo.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "api"
-version = "0.5.0"
+version = "0.5.1-SNAPSHOT"
```
[the remark this answers is not in the document]
Comment (ysibirski) [reply api/Cargo.toml:3]: Confirmed that from here: https://github.com/rust-lang/cargo/issues/2222

We should be good to go with this naming convention: https://doc.rust-lang.org/cargo/reference/semver.html

**Review by cronicc — CHANGES_REQUESTED** [review]

File: rust-toolchain [hunk 2]
```diff
@@ -1 +0,0 @@
-1.51.0
```
Comment (DanieleDiBenedetto) [inline rust-toolchain:1]: Why this has been deleted ? It is useful when an user clones and compiles this repo manually

File: .travis.yml [hunk 3]
```diff
@@ -12,21 +12,21 @@ services:
...
-    #  env: CONTAINER_RUST_VER='nightly-2021-04-25' RUSTFLAGS='-C target-feature=+bmi2,+adx --emit=asm'
-
-script: >-
-  source ci/setup_env.sh &&
-  docker build --pull --no-cache -t zencash/zendoo-sc-cryptolib-builder ./ci &&
-  bash -c "docker run --rm -v $(pwd):/build -v ${HOME}/key.asc:/key.asc --tmpfs /tmp:uid=$(id -u),gid=$(id -g),exec,mode=1777 \
-    --tmpfs /run:uid=$(id -u),gid=$(id -g),exec,mode=1777 -e RUSTFLAGS -e LOCAL_USER_ID=$(id -u) -e LOCAL_GRP_ID=$(id -g) -e CARGO_AUDIT_EXIT_ON_ERROR \
-    $(env | grep -E '^CONTAINER_' | sed -n '/^[^\t]/s/=.*//p' | sed '/^$/d' | sed 's/^/-e /g' | tr '\n' ' ') \
-    zencash/zendoo-sc-cryptolib-builder /build/ci/start_ci.sh"
+      env: IMAGE_TAG='bionic_rust-1.51.0_jdk-11_latest' RUSTFLAGS='-C target-feature=+bmi2,+adx --emit=asm' TESTS='000,001,002,003,004,005,006'
```
Comment (DanieleDiBenedetto) [inline .travis.yml:30]: Having 6 stages is a bit overdoing perhaps ?

**Review by DanieleDiBenedetto — CHANGES_REQUESTED** [review]

File: ci/devtools/travis-cli/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,19 @@
+FROM ruby:2.6-alpine
+
+RUN apk add --no-cache \
+		git
+
+ENV TRAVIS_CLI_VERSION 1.10.0
```
[the remark this answers is not in the document]
Comment (ysibirski) [reply ci/devtools/travis-cli/Dockerfile:6]: Done

File: ci/set_version.sh [hunk 5]
```diff
@@ -0,0 +1,120 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (ysibirski) [reply ci/set_version.sh:1]: Done

File: ci/set_version.sh [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+         DIGIT.DIGIT.DIGIT or DIGIT.DIGIT.DIGIT-SNAPSHOT
+
+         For example:
+         ./set_version.sh 5.5.5 5.5.5-SNAPSHOT
+         ./set_version.sh 5.5.5-SNAPSHOT 5.5.5
+BLOCK
+  fn_die "Exiting ..."
+}
+
+# Checking if cargo is installed for MAC and Linux OS
```
[the remark this answers is not in the document]
Comment (ysibirski) [reply ci/set_version.sh:39]: Implemented

File: ci/set_version.sh [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    fn_die "Fix it! The OLD version does not match with CURRENT version under ${dir}/Cargo.toml file\nCurrent version is: ${current_pom_version}"
+  fi
+
+  echo "" && echo "=== Modifying Cargo.toml file under ${dir} location ===" && echo ""
+
+  # Final version after
+  "${HOME}"/.cargo/bin/set-cargo-version "${dir}/Cargo.toml" "${version_new}"
+done
+
+# Changing crates version under caro.lock file
```
[the remark this answers is not in the document]
Comment (ysibirski) [reply ci/set_version.sh:102]: Fixed

File: ci/set_version.sh [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+if ! [[ "${version_old}" =~ ^[0-9]+\.[0-9]+\.[0-9]+(-SNAPSHOT)?$ ]]; then
+  usage
+fi
+
+if ! [[ "${version_new}" =~ ^[0-9]+\.[0-9]+\.[0-9]+(-SNAPSHOT)?$ ]]; then
+  usage
+fi
+
+# Changing version numbers under pom.xml file(s)
+for dir in "${jni_xml_location}"; do
```
[the remark this answers is not in the document]
Comment (ysibirski) [reply ci/set_version.sh:74]: Fixed

File: ci/setup_env.sh [hunk 9]
```diff
@@ -8,29 +8,56 @@ echo "TRAVIS_TAG: $TRAVIS_TAG"
...
-  gpg -v --batch --keyserver hkp://pgp.mit.edu:80 --recv-keys $MAINTAINER_KEYS
-  if git verify-tag -v "${TRAVIS_TAG}"; then
-    echo "Valid signed tag"
-    if [[ "${CONTAINER_RUST_VER}" != *"nightly"* ]]; then
-      echo "Publishing - this is a release"
-      if [ "${TRAVIS_TAG}" != "${pom_version}" ]; then
-        echo "Aborting, tag differs from the pom file."
-        exit 1
+  # shellcheck disable=SC2086
+  import_gpg_keys ${MAINTAINER_KEYS}
```
[the remark this answers is not in the document]
Comment (ysibirski) [reply ci/setup_env.sh:48]: Fixed

File: ci/setup_env.sh [hunk 10]
```diff
@@ -8,29 +8,56 @@ echo "TRAVIS_TAG: $TRAVIS_TAG"
...
+
+  if (check_signed_tag "${TRAVIS_TAG}"); then
+    if [[ "${IMAGE_TAG}" != *"nightly"* ]]; then
+      if ! [[ ${TRAVIS_TAG} =~ "${pom_version}"[0-9]*$ ]]; then
```
[the remark this answers is not in the document]
Comment (ysibirski) [reply ci/setup_env.sh:52]: I think I have copied it from sidechains-sdk. Fixed to [[ ${TRAVIS_TAG} =~ "${pom_version}"$ ]]

File: ci/docker/entrypoint_setup_gpg.sh [hunk 11]
```diff
@@ -9,4 +9,4 @@ if [ "${CONTAINER_PUBLISH}" = "true" ]; then
   gpg2 --batch --fast-import /key.asc
 fi
 
-exec "$@"
+exec "$@"
```
[the remark this answers is not in the document]
Comment (ysibirski) [reply ci/docker/entrypoint_setup_gpg.sh:12]: Fixed

File: ci/setup_env.sh [hunk 12]
```diff
@@ -8,29 +8,56 @@ echo "TRAVIS_TAG: $TRAVIS_TAG"
...
 touch "${HOME}/key.asc"
 
-if [ ! -z "${TRAVIS_TAG}" ]; then
+if [ -n "${TRAVIS_TAG}" ]; then
+  # checking if MAINTAINER_KEYS is set
+  if [ -z "${MAINTAINER_KEYS}" ]; then
+    echo "MAINTAINER_KEYS variable is not set. Make sure to set it up for release build!!!"
+  fi
+
   export GNUPGHOME="$(mktemp -d 2>/dev/null || mktemp -d -t 'GNUPGHOME')"
```
[the remark this answers is not in the document]
Comment (ysibirski) [reply ci/setup_env.sh:46]: Fixed

File: .travis.yml (hunk 3 above)
Comment (ysibirski) [reply .travis.yml:30]: It is all the commands we had before in one file but split into separate files to mimic the most recent ginger lib architecture: https://github.com/HorizenOfficial/ginger-lib/blob/master/.travis.yml

File: rust-toolchain (hunk 2 above)
Comment (ysibirski) [reply rust-toolchain:1]: Bringing this on back

File: .travis.yml (hunk 3 above)
Comment (cronicc) [reply .travis.yml:30]: This doesn't change overall execution time by much, maybe 1-2s. But has the advantage of easier identification of the last failed command should there be an error.

Comment (DanieleDiBenedetto) [reply .travis.yml:30]: I understand, but to me it's a bit overdoing having, for instance, a stage for "cargo clean" (a command that never fails and even if it does, it doesn't provide any useful information), and another stage for "cargo build" (that also won't fail if the "cargo test" stage passed)

Comment (cronicc) [reply .travis.yml:30]: If we don't care about the timing information for each command we can certainly aggregate scripts 000 - 001 or even 000 - 002 into one. On cargo audit it makes sense to have a separate script as that could fail from time to time.
```
Testscript: /build/ci/tests/000-cargo-tests.sh Status: passed Exitcode: 0 Duration: 835.152639906 seconds
Testscript: /build/ci/tests/001-cargo-clean.sh Status: passed Exitcode: 0 Duration: 0.261754856 seconds
Testscript: /build/ci/tests/002-cargo-build.sh Status: passed Exitcode: 0 Duration: 411.551721828 seconds
Testscript: /build/ci/tests/003-cargo-audit.sh Status: passed Exitcode: 0 Duration: 26.128532029 seconds
Testscript: /build/ci/tests/004-mvn-build.sh Status: passed Exitcode: 0 Duration: 24.674652258 seconds
Testscript: /build/ci/tests/005-mvn-tests.sh Status: passed Exitcode: 0 Duration: 191.408469648 seconds
Testscript: /build/ci/tests/006-mvn-publish.sh Status: passed Exitcode: 0 Duration: 0.003049580 seconds
```

**Review by cronicc — APPROVED** [review]

File: .travis.yml (hunk 3 above)
Comment (DanieleDiBenedetto) [reply .travis.yml:30]: I agree merging 000 to 002. Probably also 004-006 ?

Comment (ysibirski) [reply .travis.yml:30]: Mergin 000-002

**Pull request merged**